README
# test proxy web server written in python.

## what is it used for?
it serves a local server, which proxies requests from one domain to another.

this would allow a developer to work locally, and be able to route all requests
to `somedomain.com` to a server of their choice, eliminating the need to modify
the `/etc/hosts` file, or configuring local domain name servers.


## how does it work?
run the proxy server and navigate to the url http://localhost:3128/proxyserver.pac
file will best illustrate how the javascript-like syntax will work:
    `python server.py`


CHANGES.md
pacman changelog
================


test_proxy.py
#!/usr/bin/python

'''Test proxy web server.'''

import BaseHTTPServer, select, socket, SocketServer, urlparse
import logging
import logging.handlers
import getopt
import sys
import signal
import threading
from types import FrameType, CodeType
from time import sleep

DEFAULT_LOG_FILENAME = "proxy.log"
HOSTNAME = 'localhost'
PROXY_HOSTNAME = 'justdropt.com'
PORT = 3128
FORWARD_HOSTNAME = 'localhost'
FORWARD_PORT = 8080

class ProxyHandler(BaseHTTPServer.BaseHTTPRequestHandler):
  __base = BaseHTTPServer.BaseHTTPRequestHandler
  __base_handle = __base.handle

  def handle(self):
    (ip, port) = self.client_address
    self.server.logger.log(logging.INFO, "Request from '%s'", ip)
    if hasattr(self, 'allowed_clients') and ip not in self.allowed_clients:
      self.raw_requestline = self.rfile.readline()
      if self.parse_request(): self.send_error(403)
    else:
      self.__base_handle()

  def _connect_to(self, soc):
    host_port =(FORWARD_HOSTNAME, FORWARD_PORT)
    self.server.logger.log(logging.INFO, "connect to %s:%d", host_port[0], host_port[1])
    try:
      soc.connect(host_port)
    except socket.error, arg:
      try:
        msg = arg[1]
      except:
        msg = arg
      self.send_error(404, msg)
      return 0
    return 1

  def output_proxy(self):
    # send requests to google apps domain to direct
    self.wfile.write('''
    function FindProxyForURL(url, host) {
      if (shExpMatch(url,"*mail.%s*"))
        return "DIRECT";
      if (shExpMatch(host,"*mail.%s*"))
        return "DIRECT";
      if (shExpMatch(url,"*%s*"))
        return "PROXY %s:%d";
      return "DIRECT";
    }''' % (PROXY_HOSTNAME, PROXY_HOSTNAME, HOSTNAME, PORT))
    return

  def do_GET(self):
    (scm, netloc, path, params, query, fragment) = urlparse.urlparse(
      self.path, 'http')

    if self.path == '/proxy.pac':
      return self.output_proxy()

    if scm != 'http' or fragment or not netloc:
      self.send_error(400, "bad url %s" % self.path)
      return

    message = ''
    if "Content-Length" in self.headers:
      content_length = self.headers["Content-Length"]
      print "message length: ", content_length
      if content_length > 0:
        message = self.rfile.read(int(content_length))
        print "message: ", message

    soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
      if self._connect_to(soc):
        self.log_request()
        soc.send("%s %s %s\r\n" %(self.command,
                                  urlparse.urlunparse(('', '', path,
                                                       params, query,
                                                       '')),
                                  self.request_version))
        self.headers['Connection'] = 'close'
        del self.headers['Proxy-Connection']
        print 'URL: %s request' % self.path   
        for key_val in self.headers.items():
          soc.send("%s: %s\r\n" % key_val)
          print '%s: %s\n' % key_val
        soc.send("\r\n%s\r\n" % message)
        self._read_write(soc)
    finally:
      soc.close()
      self.connection.close()

  def _read_write(self, soc, max_idling=20, local=False):
    iw = [self.connection, soc]
    ow = []
    count = 0
    while 1:
      count += 1
      (ins, _, exs) = select.select(iw, ow, iw, 1)
      if exs:
        break
      if ins:
        for i in ins:
          if i is soc:
            out = self.connection
          else:
            out = soc
          data = i.recv(8192)
          if data:
            out.send(data)
            count = 0
      if count == max_idling: break

  do_HEAD = do_GET
  do_POST = do_GET
  do_PUT  = do_GET
  do_DELETE=do_GET

  def log_message(self, format, *args):
    self.server.logger.log(logging.INFO, "%s %s", self.address_string(),
                format % args)
    
  def log_error(self, format, *args):
    self.server.logger.log(logging.ERROR, "%s %s", self.address_string(),
                format % args)

class ThreadingHTTPServer(SocketServer.ThreadingMixIn,
               BaseHTTPServer.HTTPServer):
  def __init__(self, server_address, RequestHandlerClass, logger=None):
    BaseHTTPServer.HTTPServer.__init__(self, server_address,
                      RequestHandlerClass)
    self.logger = logger

def logSetup():
  logger = logging.getLogger("TinyHTTPProxy")
  logger.setLevel(logging.INFO)
  handler = logging.StreamHandler()
  fmt = logging.Formatter("[%(asctime)-12s.%(msecs)03d] "
               "%(levelname)-8s {%(name)s %(threadName)s}"
               " %(message)s",
               "%Y-%m-%d %H:%M:%S")
  handler.setFormatter(fmt)
    
  logger.addHandler(handler)
  return logger

def usage(msg=None):
  if msg: print msg
  print sys.argv[0], "[-p port] [-l logfile] [-dh] [allowed_client_name ...]]"
  print
  print "   -p     - Port to bind to"
  print

def handler(signo, frame):
  while frame and isinstance(frame, FrameType):
    if frame.f_code and isinstance(frame.f_code, CodeType):
      if "exit_event" in frame.f_code.co_varnames:
        frame.f_locals["exit_event"].set()
        return
    frame = frame.f_back
  
def main():
  logfile = None
  PORT = 3128
  exit_event = threading.Event()

  
  try: opts, args = getopt.getopt(sys.argv[1:], "l:dhp:", [])
  except getopt.GetoptError, e:
    usage(str(e))
    return 1

  for opt, value in opts:
    # TODO: add support for the forward host and port through the command line
    if opt == "-p":
      port = int(value)
    else:
      usage()
      return 0
    
  # setup the log file
  logger = logSetup()
  
  signal.signal(signal.SIGINT, handler)

  server_address = (HOSTNAME, PORT)
  ProxyHandler.protocol = "HTTP/1.0"
  httpd = ThreadingHTTPServer(server_address, ProxyHandler, logger)
  sa = httpd.socket.getsockname()
  print "Servering HTTP on", sa[0], "port", sa[1]
  req_count = 0
  while not exit_event.isSet():
    try:
      httpd.handle_request()
      req_count += 1
      if req_count == 1000:
        logger.log(logging.INFO, "Number of active threads: %s",
              threading.activeCount())
        req_count = 0
    except select.error, e:
      if e[0] == 4 and exit_event.isSet(): pass
      else:
        logger.log(logging.CRITICAL, "Errno: %d - %s", e[0], e[1])
  logger.log(logging.INFO, "Server shutdown")
  return 0

if __name__ == '__main__':
  sys.exit(main())



pacman.py
#!/usr/bin/env python
"""
  pacman
  ~~~~~~

  pacfile proxy web server written in python. useful for developing and testing
  cross-domain applications. one step install, easily configured via yaml.


  :copyright: (c) 2015 by gregorynicholas.
  :license: BSD, see LICENSE for more details.
"""
import sys
import logging
import logging.handlers
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
from optparse import OptionParser
from SocketServer import ThreadingMixIn
from threading import Event
from threading import activeCount
import select
import socket
from urlparse import urlparse, urlunparse
from types import FrameType, CodeType
from signal import signal, SIGINT

try:
  import yaml
except ImportError:
  print "python-yaml dependency not installed.."


USAGE = """\
pacman

pacman [OPTIONS] ...

description: a python web proxy server to nail cross-domain application
development. one of the most essential utilities for the cross-domain
application toolbelt since resurrections n' shit.

OPTIONS:
  --host=HOST           host name to bind the proxy server to.
  --port=PORT           port to bind the proxy server to.
  --max-threads         maximum number of active threads.
  --log-level           logging output level.defaulst
  --proxy-config        proxy configuration yaml file.
"""


CONFIG_PATH = "pacman.yaml"
# pylint #E501 disable
LOG_FMT = "[%(asctime)-12s] %(levelname)-8s {%(name)s \
%(threadName)s} %(message)s "


def load_proxy_config(path):
  """
  load server config definitions from a yaml file.
  """
  with open(path, 'r') as f:
    result = yaml.safe_load(f)
    f.close()
  return result


class ProxyHandler(BaseHTTPRequestHandler):
  """
  proxy server request handler.
  """

  def __init__(self, *args, **kwargs):
    BaseHTTPRequestHandler.__init__(self, *args, **kwargs)
    self.protocol = "HTTP/1.0"
    self.protocol_scheme = 'http'
    self._proxy_config = None
    self._proxy_rules = None
    self._pacfile_config = None

  @property
  def proxy_config(self):
    if not hasattr(self, '_proxy_config') or self._proxy_config is None:
      self._proxy_config = load_proxy_config(CONFIG_PATH)
    return self._proxy_config

  @property
  def proxy_rules(self):
    if not hasattr(self, '_proxy_rules') or self._proxy_rules is None:
      self._proxy_rules = self.proxy_config.get('proxy_rules')
    return self._proxy_rules

  @property
  def pacfile_config(self):
    if not hasattr(self, '_pacfile_config') or self._pacfile_config is None:
      self._pacfile_config = self.proxy_config.get('pacfile')
    return self._pacfile_config


  def handle(self):
    """
    handles all requests
    """
    (ip, port) = self.client_address
    self.server.logger.info("handling request from ip: {}".format(ip))

    if hasattr(self, 'allowed_clients') and ip not in self.allowed_clients:
      self.raw_requestline = self.rfile.readline()
      if self.parse_request():
        self.send_error(403)

    else:
      return BaseHTTPRequestHandler.handle(self)


  def render_pacfile(self):
    """
    renders the proxy pacfile from the definitions in the proxy configuration.
    """
    proxy_rule = """
      if (shExpMatch(url, "*{name}*"))
        return "PROXY {forward_host}:{forward_port}";
    """
    proxy_rules = [proxy_rule.format(**host) for host in self.proxy_rules]

    body = """
    function FindProxyForURL(url, host) {
      {}
      return "DIRECT";
    }
    """.format(''.join(proxy_rules))

    self.wfile.write(body)


  def urlparse(self, scheme, host, path, params, query, fragment):
    """
    parses the result of `urlparse()` to a dict.
    """
    return {
      'scheme': scheme,
      'host': host,
      'path': path,
      'params': params,
      'query': query,
      'fragment': fragment}


  def urlunparse(self, request):
    urlunparse(
      ('', '', request['path'], request['params'], request['query'], ''))


  def do_GET(self):
    """
    handles a GET request
    """
    request = self.urlparse(*urlparse(self.path, self.protocol_scheme))

    if self.is_pacfile_path():
      return self.render_pacfile()

    # TODO: for now, don't know how to handle fragments..
    if request['fragment']:
      self.send_error(400, "unsupported url: {}".format(self.path))
      return

    # TODO: handling other request types unimplemented..
    if request['scheme'] != 'http':
      self.send_error(400, "unsupported url: {}".format(self.path))
      return

    if not request['host']:
      self.send_error(400, "bad url: {}".format(self.path))
      return

    self.send_proxy_request(request, self.proxy_rules[request['host']])


  def is_pacfile_path(self):
    """
    returns a boolean if the current request path is the pacfile.
    """
    return self.path.lower() == self.pacfile_config['path'].lower()


  def send_proxy_request(self, request, forward_proxy):
    """
    fowards a proxy request.
    """
    body = ''
    length = self.headers.get("Content-Length", 0)
    if length > 0:
      body = self.rfile.read(int(length))
      self.server.logger.debug(
        "body (content-length {}): ".format(length, body))

    proxy_s = self.create_proxy_socket()
    try:
      if self.proxy_socket_connect(proxy_s, forward_proxy.port):
        self.log_request()

        proxy_s.send("{} {} {}\r\n".format(
          self.command, self.urlunparse(request), self.protocol))

        self.headers['Connection'] = 'close'
        del self.headers['Proxy-Connection']
        self.server.logger.debug('URL: {} request'.format(self.path))

        # send headers
        [proxy_s.send("{}: {}\r\n".format(*kv)) for kv in self.headers.items()]

        # send body
        proxy_s.send("\r\n{}\r\n".format(body))
        self.socket_rw(proxy_s)
    except Exception, e:
      self.server.logger.exception(e)

    finally:
      proxy_s.close()
      self.connection.close()


  def create_proxy_socket(self):
    return socket.socket(socket.AF_INET, socket.SOCK_STREAM)


  def proxy_socket_connect(self, proxy_s, port):
    self.server.logger.debug("connect to {}:{}".format(proxy_s, port))
    try:
      proxy_s.connect(port)

    except socket.error, arg:
      self.server.logger.exception(e)
      try:
        msg = arg[1]
      except:
        msg = arg
      self.send_error(404, msg)
      return 0

    return 1


  def socket_rw(self, proxy_s, max_idle_ticks=20):
    """
      :param proxy_s:
      :param max_idle_ticks:
      :param local:
    """
    iw = [self.connection, proxy_s]
    ow = []
    timer = 0

    while 1:
      timer += 1
      (ins, _, exs) = select.select(iw, ow, iw, 1)
      if exs:
        break

      if ins:
        for i in ins:
          if i is proxy_s:
            out = self.connection
          else:
            out = proxy_s

          data = i.recv(8192)
          if data:
            out.send(data)
            timer = 0

      if timer == max_idle_ticks:
        break


  do_PUT = do_GET
  do_HEAD = do_GET
  do_POST = do_GET
  do_DELETE = do_GET
  do_OPTIONS = do_GET


class ProxyHTTPServer(ThreadingMixIn, HTTPServer):
  """
  """

  def __init__(self, server_address, request_handler_class):
    HTTPServer.__init__(self, server_address, request_handler_class)
    self.logger = self.create_logger()

  def create_logger(self):
    rv = logging.getLogger(self.__class__.__name__.split('.')[-1])
    rv.setLevel(logging.INFO)
    handler = logging.StreamHandler()
    formatter = logging.Formatter(LOG_FMT)
    handler.setFormatter(formatter)
    rv.addHandler(handler)
    return rv


class ProxyContext(object):
  """
  `ProxyHTTPServer` context object. runs an instance of a `ProxyHTTPServer`.
  manages threading & binds to command line exit sigals.
  """

  def __init__(self, host, port, max_threads):
    """
      :param host: string hostname to bind the proxy server to.
      :param port: port integer to bind the proxy server to.
      :param max_threads: maximum number of active threads.
    """
    self.host = host
    self.port = port
    self.max_threads = max_threads
    self.active_thread_count = 0
    self.handler_class = ProxyHandler
    self.set_sigint_handler()


  @property
  def hostname(self):
    return (self.host, self.port)


  def run(self):
    """
    runs the `ProxyHTTPServer`
    """
    self.server = ProxyHTTPServer(self.hostname, self.handler_class)

    host, port = self.server.socket.getsockname()[:2]
    self.server.logger.info(
      "serving http on host: {} port: {}".format(host, port))

    self.active_thread_count = 0

    while not self.sigint_event.isSet():
      try:
        self.server.handle_request()
        self.active_thread_count = activeCount()
        self.server.logger.info(
          "active thread count: {}".format(self.active_thread_count))

        if self.active_thread_count >= self.max_threads:
          self.on_max_threads()

      except select.error, err:
        self.server.logger.debug("select.error: {}".format(e))
        self.on_sigint(*err)

    self.server.logger.info("proxy server shutdown..")


  def on_max_threads(self):
    self.server.logger.warn(
      "maximum active threads reached. resetting count..")
    # todo: more action to take here?
    self.active_thread_count = 0


  def set_sigint_handler(self):
    """
    binds a signal to listen for `KeyboardInterrupt` events.
    """
    self.sigint_event = Event()
    signal(SIGINT, self.sigint_handler)


  def sigint_handler(self, sigint, frame):
    """
    event handler method for `KeyboardInterrupt` events.

      :param sigint:
      :param frame:
    """
    self.server.logger.debug(
      "SIGINT event received: {}, {}".format(sigint, frame))

    while frame and isinstance(frame, FrameType):
      if frame.f_code and isinstance(frame.f_code, CodeType):
        # this goes through the stack to find the instance of the ProxyContext
        # to respond to this event..
        if "self" in frame.f_code.co_varnames:
          _self = frame.f_locals["self"]
          if isinstance(_self, ProxyContext):
            _self.sigint_event.set()
            exit(0)
      frame = frame.f_back


  def on_sigint(self, code, msg, *args):
    """
      :param code:
      :param msg:
    """
    self.server.logger.debug("on_sigint, code: {} msg: {}".format(code, msg))

    if code != 4 and not self.sigint_event.isSet():
      self.server.logger.critical("on_sigint error, code: {}, {}".format(code, msg))
      self.fail(msg)


def fail(message, *args):
  print >> sys.stderr, 'error:', message.format(*args),
  exit(1)


def usage():
  print >> sys.stdout, USAGE,


parser = OptionParser(
  description="""description: a python web proxy server to nail cross-domain
application development. one of the most essential utilities for the
cross-domain application toolbelt since resurrections n' shit.""",
  prog="pacman",
  usage="%prog [OPTIONS]")

parser.add_option(
  "--host", dest="host", default="localhost",
  help="host name to bind the proxy server to.")

parser.add_option(
  "--port", dest="port", default=3128,
  help="port to bind the proxy server to.")

parser.add_option(
  "--max-threads", dest="max_threads", default=1000,
  help="maximum number of active threads.")

parser.add_option(
  "--proxy-config", dest="proxy_config", default="pacman.yaml",
  help="proxy pacfile configuration yaml file.")

parser.add_option(
  "--log-level", dest="log_level", metavar="LEVEL",
  help="logging output level.")


def main():
  (options, args) = parser.parse_args()

  ProxyContext(
    options.host, options.port, options.max_threads).run()
  exit(0)


if __name__ == '__main__':
  main()


server.py
#!/usr/bin/env python
'''Test proxy web server written in python. Useful for developing for and
testing domains and sub domains.'''

import sys
import yaml
import getopt
import logging
import logging.handlers
import threading
import BaseHTTPServer, SocketServer
import select, socket, urlparse
from types import FrameType, CodeType
from signal import signal, SIGINT

PROXIES_PATH = 'proxies.yaml'

def get_servers(path):
  f = open(path, 'r')
  result = yaml.safe_load(f)
  f.close()
  return result

class ProxyHandler(BaseHTTPServer.BaseHTTPRequestHandler):
  __base = BaseHTTPServer.BaseHTTPRequestHandler
  __base_handle = __base.handle
  _servers = None
  _proxy_hosts = None
  _proxy_config = None

  @property
  def servers(self):
    if self._servers is None:
      self._servers = get_servers(PROXIES_PATH)
    return self._servers

  @property
  def proxy_hosts(self):
    if self._proxy_hosts is None:
      self._proxy_hosts = self.servers.get('proxies')
    return self._proxy_hosts

  @property
  def proxy_config(self):
    if self._proxy_config is None:
      self._proxy_config = self.servers.get('proxyserver')
    return self._proxy_config

  def handle(self):
    (ip, port) = self.client_address
    self.server.logger.info("Request from '%s'", ip)
    if hasattr(self, 'allowed_clients') and ip not in self.allowed_clients:
      self.raw_requestline = self.rfile.readline()
      if self.parse_request(): self.send_error(403)
    else:
      self.__base_handle()

  def _connect_to(self, host, port, soc):
    self.server.logger("connect to %s:%d", host, port)
    try:
      soc.connect(port)
    except socket.error, arg:
      try:
        msg = arg[1]
      except:
        msg = arg
      self.send_error(404, msg)
      return 0
    return 1

  def write_proxy_pac(self):
    result = '''
    function FindProxyForURL(url, host) {
    '''
    for host in self.proxy_hosts:
      result += '''
      if (shExpMatch(url,"*{name}*"))
        return "PROXY {forward_host}:{forward_port}";
      '''.format(**host)
    result += '''
      return "DIRECT";
    }'''
    self.wfile.write(result)

  def do_GET(self):
    (scm, netloc, path, params, query, fragment) = urlparse.urlparse(
      self.path, 'http')
    if self.path == self.proxy_config.get('path'):
      return self.write_proxy_pac()
    if scm != 'http' or fragment or not netloc:
      self.send_error(400, "bad url %s" % self.path)
      return
    server_host = self.proxy_hosts.get(netloc)
    message = ''
    if "Content-Length" in self.headers:
      content_length = self.headers["Content-Length"]
      print "message length: ", content_length
      if content_length > 0:
        message = self.rfile.read(int(content_length))
        print "message: ", message

    soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
      if self._connect_to(server_host.name, server_host.port, soc):
        self.log_request()
        soc.send("%s %s %s\r\n" %(
          self.command, urlparse.urlunparse(('', '', path, params, query, '')),
          self.request_version))
        self.headers['Connection'] = 'close'
        del self.headers['Proxy-Connection']
        # print 'URL: %s request' % self.path
        for key_val in self.headers.items():
          soc.send("%s: %s\r\n" % key_val)
          print '%s: %s\n' % key_val
        soc.send("\r\n%s\r\n" % message)
        self._read_write(soc)
    finally:
      soc.close()
      self.connection.close()

  def _read_write(self, soc, max_idling=20, local=False):
    iw = [self.connection, soc]
    ow = []
    count = 0
    while 1:
      count += 1
      (ins, _, exs) = select.select(iw, ow, iw, 1)
      if exs:
        break
      if ins:
        for i in ins:
          if i is soc:
            out = self.connection
          else:
            out = soc
          data = i.recv(8192)
          if data:
            out.send(data)
            count = 0
      if count == max_idling: break

  do_PUT     = do_GET
  do_HEAD    = do_GET
  do_POST    = do_GET
  do_DELETE  = do_GET
  do_OPTIONS = do_GET

  def log_message(self, format, *args):
    self.server.logger.info("%s %s", self.address_string(), format % args)

  def log_error(self, format, *args):
    self.server.logger.error("%s %s", self.address_string(), format % args)


class ThreadingHTTPServer(SocketServer.ThreadingMixIn, BaseHTTPServer.HTTPServer):
  def __init__(self, server_address, request_handler_class):
    BaseHTTPServer.HTTPServer.__init__(
      self, server_address, request_handler_class)
    self.logger = self.create_logger()

  def create_logger(self):
    result = logging.getLogger("TinyHTTPProxy")
    result.setLevel(logging.INFO)
    hldr = logging.StreamHandler()
    format = logging.Formatter("[%(asctime)-12s.%(msecs)03d] "
      "%(levelname)-8s {%(name)s %(threadName)s}"
      " %(message)s",
      "%Y-%m-%d %H:%M:%S")
    hldr.setFormatter(format)
    result.addHandler(hldr)
    return result


def usage(msg=None):
  if msg: print msg
  print sys.argv[0], "[-p port] [-l logfile] [-dh] [allowed_client_name ...]]"
  print
  print "   -p     - Port to bind to"
  print


def handler(signo, frame):
  while frame and isinstance(frame, FrameType):
    if frame.f_code and isinstance(frame.f_code, CodeType):
      if "exit_event" in frame.f_code.co_varnames:
        frame.f_locals["exit_event"].set()
        return
    frame = frame.f_back


def main():
  PORT = 3128
  HOST = 'localhost'
  exit_event = threading.Event()

  try:
    opts, args = getopt.getopt(sys.argv[1:], "l:dhp:", [])
  except getopt.GetoptError, e:
    usage(str(e))
    return 1

  for opt, value in opts:
    # TODO: add support for the forward host and port through the command line
    if opt == "-p":
      PORT = int(value)
    else:
      usage()
      return 0

  signal(SIGINT, handler)
  server_address = (HOST, PORT)
  ProxyHandler.protocol = "HTTP/1.0"
  httpd = ThreadingHTTPServer(server_address, ProxyHandler)
  sockname = httpd.socket.getsockname()
  httpd.logger.info("Serving HTTP on: %s port: %s", sockname[0], sockname[1])
  active_threads_count = 0
  max_active_threads_count = 1000
  while not exit_event.isSet():
    try:
      httpd.handle_request()
      active_threads_count += 1
      if active_threads_count >= max_active_threads_count:
        httpd.logger.info("Number of active threads: %s",
          threading.activeCount())
        active_threads_count = 0
    except select.error, err:
      if err[0] == 4 and exit_event.isSet():
        pass
      else:
        httpd.logger.critical("Errno: %d - %s", err[0], err[1])
  httpd.logger.info("Server shutdown..")
  return 0


if __name__ == '__main__':
  sys.exit(main())
